src/coreclr/jit/morph.cpp

```diff
@@ -13780,10 +13780,80 @@ void Compiler::fgMorphBlock(BasicBlock* block)
     {
-        // For now, each block starts with an empty table, and no available assertions
-        //
-        optAssertionReset(0);
-        apLocal = BitVecOps::MakeEmpty(apTraits);
+        if (!optCrossBlockLocalAssertionProp)
+        {
+            // Each block starts with an empty table, and no available assertions
+            //
+            optAssertionReset(0);
+            apLocal = BitVecOps::MakeEmpty(apTraits);
+        }
+        else
+        {
+            // Determine if this block can leverage assertions from its pred blocks.
+            //
+            // Some blocks are ineligible.
+            //
+            bool canUsePredAssertions = ((block->bbFlags & BBF_CAN_ADD_PRED) == 0) && !bbIsHandlerBeg(block);
+
+            // Validate all preds have valid info
+            //
+            if (canUsePredAssertions)
+            {
+                bool hasPred = false;
+
+                for (BasicBlock* const pred : block->PredBlocks())
+                {
+                    // A higher postorder number means the block appears earlier in
+                    // the postorder. Use this to detect if a pred's assertion info is available.
+                    //
+                    if (pred->bbPostorderNum > block->bbPostorderNum)
+                    {
+                        // Yes, available. If this is the first pred, copy.
+                        // If this is a subsequent pred, intersect.
+                        //
+                        if (!hasPred)
+                        {
+                            apLocal = BitVecOps::MakeCopy(apTraits, pred->bbAssertionOut);
+                            hasPred = true;
+                        }
+                        else
+                        {
+                            BitVecOps::IntersectionD(apTraits, apLocal, pred->bbAssertionOut);
+                        }
+
+                        continue;
+                    }
+
+                    // No, not available.
+                    //
+                    JITDUMP(FMT_BB " pred " FMT_BB " not processed; clearing assertions in\n", block->bbNum,
+                            pred->bbNum);
+                    canUsePredAssertions = false;
+                }
+
+                // If there were no preds, there are no asserions in.
+                //
+                if (!hasPred)
+                {
+                    JITDUMP(FMT_BB " has no preds, so no assertions in\n", block->bbNum);
+                    canUsePredAssertions = false;
+                }
+
+                if (canUsePredAssertions)
+                {
+                    JITDUMPEXEC(optDumpAssertionIndices("Assertions in: ", apLocal));
+                }
+            }
+            else
+            {
+                JITDUMP(FMT_BB " ineligible for cross-block; clearing assertions in\n", block->bbNum);
+            }
+
+            if (!canUsePredAssertions)
+            {
+                apLocal = BitVecOps::MakeEmpty(apTraits);
```

Comment:
Does this also need `optAssertionReset(0);`?

Can the `(!optCrossBlockLocalAssertionProp)` and the "failure" cases be combined (tail merged)?

Reply:
> Does this also need `optAssertionReset(0);`?

No. There are 3 modes of operation here:
1. classic: block starts with empty table via `Reset(0)` and empty BV
2. cross-block: block starts with current table and intersection of pred BVs (which may end up empty)
3. cross-block special case: block starts with current table and empty BV

Modes 2 and 3 can mix with each other, but not with mode 1.

We use mode 1 if the feature is disabled, or if the method has a lot of locals.

When the feature is enabled, and the method does not have a lot of locals, we use mode 2 (if `canUsePredAssertions`) or mode 3 (otherwise) depending on the various block properties.

> Can the `(!optCrossBlockLocalAssertionProp)` and the "failure" cases be combined (tail merged)?

Let me look at tweaking this.